src/lib.rs

```diff
@@ -211,3 +220,189 @@ impl fmt::Display for Error {
+{
+    actor.started(&mut mailbox).await;
+
+    while let ControlFlow::Continue(()) = yield_once(&mut mailbox, &mut actor).await {}
+
+    actor.stopped().await
+}
+
+/// Yield one message from the [`Mailbox`] and process it using the given actor.
+pub async fn yield_once<A>(mailbox: &mut Mailbox<A>, actor: &mut A) -> ControlFlow<(), ()>
+where
+    A: Actor,
+{
+    let message = mailbox.next().await;
+
+    tick(message, actor, mailbox).await
+}
+
+/// Process exactly on message from the mailbox on the given actor.
+pub fn tick<'a, A>(
+    message: Message<A>,
+    actor: &'a mut A,
+    mailbox: &'a mut Mailbox<A>,
+) -> TickFuture<'a, A>
+where
+    A: Actor,
+{
+    TickFuture::new(message.0, actor, mailbox)
+}
+
+/// Handle any incoming messages for the actor while running a given future. This is similar to
+/// [`Context::join`], but will exit if the actor is stopped, returning the future. Returns
+/// `Ok` with the result of the future if it was successfully completed, or `Err` with the
+/// future if the actor was stopped before it could complete. It is analagous to
+/// [`futures::select`](futures_util::future::select).
+///
+/// ## Example
+///
+/// ```rust
+/// # use std::time::Duration;
+/// use futures_util::future::Either;
+/// # use xtra::prelude::*;
+/// # use smol::future;
+/// # struct MyActor;
+/// # #[async_trait] impl Actor for MyActor { type Stop = (); async fn stopped(self) {} }
+///
+/// struct Stop;
+/// struct Selecting;
+///
+/// #[async_trait]
+/// impl Handler<Stop> for MyActor {
+///     type Return = ();
+///
+///     async fn handle(&mut self, _msg: Stop, ctx: &mut Context<Self>) {
+///         ctx.stop_self();
+///     }
+/// }
+///
+/// #[async_trait]
+/// impl Handler<Selecting> for MyActor {
+///     type Return = bool;
+///
+///     async fn handle(&mut self, _msg: Selecting, ctx: &mut Context<Self>) -> bool {
+///         // Actor is still running, so this will return Either::Left
+///         match xtra::select(ctx.mailbox(), self, future::ready(1)).await {
+///             Either::Left(ans) => println!("Answer is: {}", ans),
+///             Either::Right(_) => panic!("How did we get here?"),
+///         };
+///
+///         let addr = ctx.mailbox().address();
+///         let select = xtra::select(ctx.mailbox(), self, future::pending::<()>());
+///         let _ = addr.send(Stop).split_receiver().await;
+///
+///         // Actor is stopping, so this will return Err, even though the future will
+///         // usually never complete.
+///         matches!(select.await, Either::Right(_))
+///     }
+/// }
+///
+/// # #[cfg(feature = "with-smol-1")]
+/// # smol::block_on(async {
+/// let addr = MyActor.create(None).spawn(&mut xtra::spawn::Smol::Global);
+/// assert!(addr.is_connected());
+/// assert_eq!(addr.send(Selecting).await, Ok(true)); // Assert that the select did end early
+/// # })
+///
+/// ```
+pub async fn select<A, F, R>(mailbox: &mut Mailbox<A>, actor: &mut A, mut fut: F) -> Either<R, F>
```

Comment:
Is there a reason to have this as a free function rather than on `Mailbox` itself?

Reply:
I figured it doesn't really fit on `Mailbox`. Why not on `Actor`? Why not an extension trait for `Future`s? I don't see how any of these types are "privileged" in where this code should live so a free function felt like the right thing to do.

Reply:
I guess. Personally I feel like it fits on mailbox since you are selecting between mailbox messages and the given future. That being said, though, you are also selecting on handling those messages

Reply:
The 2nd reason is a taste thing. Rust is not an OOP language and the main reason for functions to live _on_ a type is for information hiding purposes where the function needs to access internal state. This is not the case here so it can just be a free function.

Reply:
I just drafted this up as quickly as possible, the order of the parameters should probably be thought about a bit more :)

Reply:
What is your latest position on this? Would you like this function to be on `Mailbox` or can I move forward on the free function based design?

Reply:
I think I mentioned this elsewhere but I agree that a free function based design makes sense here